src/backend/optimizer/plan/createplan.c

```diff
@@ -784,9 +784,12 @@ create_join_plan(PlannerInfo *root, JoinPath *best_path)
 
 	/* CDB: if the join's locus is bottleneck which means the
 	 * join gang only contains one process, so there is no
-	 * risk for motion deadlock.
+	 * risk for motion deadlock. But prefetch_inner may still be necessary in the
+	 * case of HashJoin with SharedScan, the writing part of which is always in
+	 * the inner. So do not override it in the case of HashJoin.
 	 */
-	if (CdbPathLocus_IsBottleneck(best_path->path.locus))
+	if (CdbPathLocus_IsBottleneck(best_path->path.locus) &&
+		!(root->config->gp_cte_sharing && IsA(plan, HashJoin)))
```

Comment:
Why didn't you add a check for `root->config->gp_cte_sharing` to the previous solution with changes to `create_nestloop_plan` and `create_mergejoin_plan`?

Reply:
I don't see any point in moving logic to join creation functions if it will be applied to all types with a single exception.

Reply:
Then again the [question](https://github.com/arenadata/gpdb/pull/620/commits/7dd69e6b1500716e82f3d04579c7a214ab084515#r1338483446) remains valid.

Reply:
maybe move it higher in the same function with the removal of prefetch_joinqual and prefetch_qual?